### PR TITLE
Updating Documentation Binaries for Correct Minor Versions

### PR DESCRIPTION
**Pull request by AMollis**
# Pull Request Description: Updating Documentation Binaries for Correct Minor Versions

## Overview 

The current issue lies in the incorrect minor versions of our documentation binaries. Specifically, the `AssemblyInfo.cs` files are not being updated with the latest major/minor version during the documentation build pipeline. To address this, we propose applying the same versioning logic used for package creation to the documentation generation process.

Here are the key changes:

1. **Consolidating Versioning Logic:** We’ve consolidated the versioning logic from both `pack-upm.ps1` and `repackage-for-release.ps1` into a single PowerShell script named `update-versions.ps1`. This script can handle versioning for both prerelease and release packages, following the current version patterns (semantic versioning with labels and appended revision numbers as needed). The goal here is to not change the versioning logic. In the future we may want to revisit this. For example, in a future pull request, we might want to remove the usage of the `pre.x` tags on experimental packages, where `x` is an additional experimentation revision number shared across all package. This experimental tag adds additional complexity to the build scripts, and might be avoided if we just used a `-prerelease` label with the already utilized semantic versioning.
3. **Handling Prerelease to Release Transition:** The new script is also capable of re-versioning a prerelease package to a release package seamlessly.
4. **Integration with `doc.yaml`**: We’ve integrated the new script into `doc.yaml`. Now, when generating documentation binaries, the script ensures that they appear as release packages. This adjustment allows for correct moniker (i.e., directory) names to be created. That is, when a moniker is created for a package, the external doc generation script queries the major and minor versions of the doc libraries (DLL).
5. **Improved Version Consistency:** Previously, the doc generation script relied on hard-coded major and minor versions, which didn’t always match the versions specified in the packages’ `package.json` files. With this change, we maintain a single source of versioning truth—the `package.json`—and expect the build pipeline to propagate the version to other relevant areas, including the AssemblyInfo.cs files.

By implementing these modifications, we aim to enhance the accuracy and consistency of our documentation binaries. 

## Fixes
- https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/637

### CODE REVIEW

File: Pipelines/Scripts/pack-upm.ps1 [hunk 1]
```diff
@@ -10,61 +10,54 @@
...
+.PARAMETER ReleaseLabel
+    The tag to append after the version (e.g. "internal" or "prerelease"). Leave blank for a release build.
+.PARAMETER ExperimentLabel
+    An additional tag to append after the version, to append after the release label (e.g. "pre.1"). Historically used for the MRTK3 packages that are still experimental.
 .PARAMETER Revision
-    The revision number for the build.  (e.g. "230508.3" or "2") 
+    The revision number for the build, to append after the release labal and suffix.
 .PARAMETER ReleasePackages
-    An array of the package names that are approved for release.  If the package isn't in this array, it will get labeled -Pre
+    An array of the package names that are no longer  If the package isn't in this array, it will get labeled with the ExperimentLabel.
```
Comment (shaynie) [inline Pipelines/Scripts/pack-upm.ps1:22]: Complete sentence?

File: Pipelines/Scripts/pack-upm.ps1 [hunk 2]
```diff
@@ -10,61 +10,54 @@
...
     New-Item $OutputDirectory -ItemType Directory | Out-Null
 }
 
 $OutputDirectory = Resolve-Path -Path $OutputDirectory
-Write-Output "OutputDirectory: $OutputDirectory"
 
-Write-Output "Release packages: $releasePkgs"
+Write-Host ""
+Write-Host -ForegroundColor Blue "======================================="  
+Write-Host -ForegroundColor Blue "Packing All Pacakges"
```
Comment (shaynie) [inline Pipelines/Scripts/pack-upm.ps1:50]: spelling Packages

File: Pipelines/Scripts/repackage-for-release.ps1 [hunk 3]
```diff
@@ -14,26 +14,28 @@
...
     New-Item $OutputDirectory -ItemType Directory | Out-Null
 }
 
 $OutputDirectory = Resolve-Path -Path $OutputDirectory
-Write-Output "OutputDirectory: $OutputDirectory"
 
-Write-Output "Release packages: $releasePkgs"
+Write-Host ""
+Write-Host -ForegroundColor Blue "======================================="  
+Write-Host -ForegroundColor Blue "Repackaging All Pacakges For Release"
```
Comment (shaynie) [inline Pipelines/Scripts/repackage-for-release.ps1:30]: spelling

File: Pipelines/Scripts/repackage-for-release.ps1 [hunk 4]
```diff
@@ -105,8 +102,13 @@ try {
         Write-Output "Packing $packageFriendlyName to $OutputDirectory"
         npm pack $packagePath -pack-destination $OutputDirectory
     }
+
+    Write-Host ""
+    Write-Host -ForegroundColor Blue "======================================="  
+    Write-Host -ForegroundColor Blue "Successfully Packed Release Pacakges"
```
Comment (shaynie) [inline Pipelines/Scripts/repackage-for-release.ps1:108]: spelling

File: Pipelines/Scripts/update-versions.ps1 [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    if (-not $packageName) {
+        return # this is not an MRTK package, so skip
+    }
+
+    $packageName = $packageName.Matches[0].Value
+    $packagePath = $_.Directory
+
+    Write-Host ""
+    Write-Host -ForegroundColor Green "======================================="  
+    Write-Host -ForegroundColor Green "Updating Pacakge Version"
```
Comment (shaynie) [inline Pipelines/Scripts/update-versions.ps1:80]: spelling

File: Pipelines/Scripts/update-versions.ps1 [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+    Write-Output "Patching CHANGELOG.md version to [$version$label] - $year-$month-$day"
+    Get-ChildItem -Path $packagePath/CHANGELOG.md -Recurse | ForEach-Object {            
+        (Get-Content -Path $_ -Raw) -Replace "## \[$version(-[a-zA-Z0-9.]+)?\] - \b\d{4}\b-\b(0[1-9]|1[0-2])\b-\b(0[1-9]|[12][0-9]|3[01])\b", "## [$version$label] - $year-$month-$day" | Set-Content -Path $_ -NoNewline
+    }
+}
+
+Write-Host ""
+Write-Host -ForegroundColor Green "=======================================" 
+Write-Host -ForegroundColor Green "Successfully Updated Packages Versions"
```
Comment (shaynie) [inline Pipelines/Scripts/update-versions.ps1:141]: Package singular

**Review by shaynie — APPROVED** [review]
Minor nits - spelling
Approving without running the new scripts

File: Pipelines/Scripts/pack-upm.ps1 (hunk 1 above)
Comment (AMollis) [reply Pipelines/Scripts/pack-upm.ps1:22]: fix in next commit

**Review by shaynie — APPROVED** [review]

**Pull request merged**